### PR TITLE
Improve Dedent behavior

### PR DESCRIPTION
**Pull request by WindSoilder**
handle the remaining of #1046 :  Hitting o to go to a new line auto indents. If you then immediately hit Esc, it should delete all of the whitespace and just leave a blank line.

### CODE REVIEW

**Comment by archseer** [comment]
Note that since you used a single commit for both changes I had to revert the kill_to_line_end change too, feel free to add it back (but keep it in a separate commit)

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -660,6 +660,17 @@ impl EditorView {
         let key_result = self.keymaps.get_mut(&mode).unwrap().get(event);
         self.autoinfo = key_result.sticky.map(|node| node.infobox());
 
+        // may cancel restore indent.
+        match &key_result.kind {
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:664]: Hmm I still don't like having this workaround inside key handling. Another option would be to check doc.changes() inside normal_mode, if it only contains a newline + indent

Comment (WindSoilder) [reply helix-term/src/ui/editor.rs:664]: Yeah, I agree that put to this location seems ugly...  I've checkout `doc.changes()`, but what if user open new line, then move cursor up/down, then press `esc` to goes into normal mode.  The doc doesn't change.

Comment (archseer) [reply helix-term/src/ui/editor.rs:664]: I think that should be expected (and users generally shouldn't move around in insert mode).

Comment (WindSoilder) [reply helix-term/src/ui/editor.rs]: Done :-)  Check the dedent inside normal mode;

And for now, when user open a new line(whatever from `enter key` or from `open above/below` commands, normal mode will auto dedent(even user enter whitespaces after open a new line)

Personaly this behavior is better than I expected.  Can you check if this behavior is feasible?

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -3594,6 +3594,39 @@ fn normal_mode(cx: &mut Context) {
...
+
+    // Will remove whitespaces when user opens a new line(may additional insert only tabs or spaces)
+    // and goto normal mode.
+    //
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
+    if doc_changes.len() == 3 {
+        if let helix_core::Operation::Retain(move_pos) = doc_changes[0] {
+            if let helix_core::Operation::Insert(ref inserted_str) = doc_changes[1] {
```
Comment (pickfire) [inline helix-term/src/commands.rs:3610]: Why not `if let [Retain(move_pos), Insert(inserted_str), _] = doc_changes {` directly?

Comment (WindSoilder) [reply helix-term/src/commands.rs]: Done.  It's simply because I don't remember the pattern..

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -3594,6 +3594,38 @@ fn normal_mode(cx: &mut Context) {
...
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
+    use helix_core::Operation;
+    if let [Operation::Retain(move_pos), Operation::Insert(ref inserted_str), Operation::Retain(_)] =
+        doc_changes
+    {
+        if move_pos + inserted_str.len32() as usize == pos
+            && inserted_str.starts_with('\n')
+            && inserted_str.chars().all(|ch| ch.is_whitespace())
```
Comment (archseer) [inline helix-term/src/commands.rs:3614]: Needs to use https://github.com/helix-editor/helix/blob/92c2d5d3bffb52a165fad75b6db2120de37b5ae4/helix-core/src/chars.rs#L35-L62

Comment (WindSoilder) [reply helix-term/src/commands.rs]: Done

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -3610,6 +3610,39 @@ fn normal_mode(cx: &mut Context) {
...
+    }
+
+    if can_restore_indent {
+        let transaction =
+            Transaction::change_by_selection(doc.text(), doc.selection(view.id), |range| {
+                let line_start_pos = text.line_to_char(range.cursor_line(text));
+                (line_start_pos, pos, None)
+            });
+        doc.apply(&transaction, view.id);
+    }
```
Comment (archseer) [inline helix-term/src/commands.rs:3644]: Let's extract this in a helper function so the code is self contained.

File: helix-term/src/commands.rs [hunk 5]
```diff
@@ -3610,6 +3610,39 @@ fn normal_mode(cx: &mut Context) {
...
+    let text = doc.text().slice(..);
+    let pos = doc.selection(view.id).primary().cursor(text);
+    let mut can_restore_indent = false;
+
+    // Will remove whitespaces when user opens a new line(may additional insert only tabs or spaces)
+    // and goto normal mode.
+    //
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
```
Comment (archseer) [inline helix-term/src/commands.rs:3623]:
```suggestion
    // Removes trailing whitespace if insert mode is exited after starting a blank new line.
```

**Review by archseer — CHANGES_REQUESTED** [review]
Tested locally and it seems okay, just two minor changes

File: helix-term/src/commands.rs (hunk 4 above)
Comment (WindSoilder) [reply helix-term/src/commands.rs]: Thank you, that make sence a lot :-) Done on that

**Pull request merged**